### PR TITLE
Psalm integration

### LINKED ISSUE #56: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [ ] Create a `.psalm.xml.dist` in the project root
- [ ] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [ ] Run `$ composer require vimeo/psalm`
- [ ] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [ ] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [ ] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [ ] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [ ] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by konarshankar07**
Signed-off-by: Shankar Konar <konar.shankar2013@gmail.com>

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

Psalm integration added

### CODE REVIEW

**Comment by boesing** [comment]
Thanks, @konarshankar07 
We have to add the new CI pipeline to this package before we can merge this.
Anyways, thanks for your efforts! This will still count for hacktoberfest due to the `hacktoberfest-accepted` label 👍🏼 

@samsonasik might handle the next steps of this, he might have plans on how to proceed with this component.

**Comment by konarshankar07** [comment]
Hello @boesing 
If you provide the steps to add the new CI pipeline then I can implement that too. I guess you are going to implement the github actions for the CI pipeline
Thanks

**Comment by froschdesign** [comment]
@boesing 
The CI workflow is already there, see at the 2.12.x branch:

> This branch is 30 commits ahead, 1 commit behind 3.0.x.

https://github.com/laminas/laminas-view/blob/2.12.x/.github/workflows/continuous-integration.yml

Some cleanup is needed: https://github.com/laminas/laminas-view/issues/71

**Comment by froschdesign** [comment]
@konarshankar07 
Please use the 2.12.x as target branch for this pull request. Thanks!

**Comment by Ocramius** [comment]
Seems like our own deprecations are at fault with the failure here :|

**Comment by boesing** [comment]
> Seems like our own deprecations are at fault with the failure here :|

actually this is due to the phpunit change regarding `deprecationsToExceptions`.
The Tests which fail are to ensure the the deprecation will be triggered (just not that well implemented) and these now fail in latest phpunit.

**Comment by konarshankar07** [comment]
@boesing @Ocramius 
Yes, I have fixed this here https://github.com/laminas/laminas-crypt/pull/14/commits/034976ef0b3097ac88df8487f7cffb0d5a531721 already and will do the same

File: .psalm.xml.dist [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    <projectFiles>
+        <directory name="bin"/>
+        <directory name="src"/>
+        <directory name="test"/>
+        <ignoreFiles>
+            <directory name="vendor"/>
+        </ignoreFiles>
+    </projectFiles>
+
+    <issueHandlers>
```
Comment (boesing) [inline .psalm.xml.dist]: if you add `psalm/plugin-phpunit` and enable it with `vendor/bin/psalm-plugin`, you should be able to remove these issue handlers.

File: .psalm.xml.dist [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0"?>
+<psalm
+    totallyTyped="true"
```
Comment (boesing) [inline .psalm.xml.dist]: `totallyTyped=true` is deprecated in favor of `errorLevel=1`.

**Review by boesing — CHANGES_REQUESTED** [review]
Overall, LGTM - just a few adjustments so we can avoid suppressions.

Another thing is, that it seems that we only have less than 20 errors in this projects. Would you mind fixing these so we end up having 100% typed code rather than suppressing these few errors via baseline? That would really be appreciated 🙏🏼

File: .psalm.xml.dist (hunk 2 above)
Comment (konarshankar07) [reply .psalm.xml.dist]: If I run the pslam with errorLevel=1 then I'm getting 4599 errors. Do we need to fix all the errors? or Am I missing something?
image
Thanks

Comment (weierophinney) [reply .psalm.xml.dist]: The idea is to establish an initial baseline; see #56 as it details this.

I've done that and pushed changes to the branch at this time.

**Pull request merged**